File: ios/GRIDDOWN/Services/SupplyAlerts.swift
import Foundation

enum ExpirationStatus {
    case expired, soon, ok, none
}

/// Mirrors the Expo app's supply alert helpers: expiration parsing,
/// low-stock detection, and an aggregated inventory alert summary.
enum SupplyAlerts {
    static let soonWindowDays = 30

    /// Parses "YYYY-MM" (treated as end of month), "YYYY-MM-DD", or ISO 8601 strings.
    static func parseExpirationDate(_ raw: String?) -> Date? {
        guard let raw else { return nil }
        let trimmed = raw.trimmingCharacters(in: .whitespaces)
        guard !trimmed.isEmpty else { return nil }

        if trimmed.contains("T") {
            return ISO8601DateFormatter().date(from: trimmed)
        }

        let parts = trimmed.split(separator: "-")
        let calendar = Calendar(identifier: .gregorian)

        if parts.count == 2,
           let year = Int(parts[0]), let month = Int(parts[1]),
           (1...12).contains(month) {
            var comps = DateComponents()
            comps.year = year
            comps.month = month + 1
            comps.day = 1
            guard let firstOfNext = calendar.date(from: comps) else { return nil }
            return calendar.date(byAdding: .second, value: -1, to: firstOfNext)
        }

        if parts.count == 3,
           let year = Int(parts[0]), let month = Int(parts[1]), let day = Int(parts[2]),
           (1...12).contains(month), (1...31).contains(day) {
            var comps = DateComponents()
            comps.year = year
            comps.month = month
            comps.day = day
            comps.hour = 23
            comps.minute = 59
            comps.second = 59
            return calendar.date(from: comps)
        }

        return ISO8601DateFormatter().date(from: trimmed)
    }

    static func isLow(_ item: SupplyItem) -> Bool {
        item.quantity <= item.minimumQuantity
    }

    static func expirationStatus(for item: SupplyItem, now: Date = Date()) -> ExpirationStatus {
        guard let date = parseExpirationDate(item.expirationDate) else { return .none }
        if date < now { return .expired }
        if let soonCutoff = Calendar.current.date(byAdding: .day, value: soonWindowDays, to: now),
           date <= soonCutoff {
            return .soon
        }
        return .ok
    }

    struct InventorySummary {
        var low: [SupplyItem] = []
        var expired: [SupplyItem] = []
        var expiringSoon: [SupplyItem] = []

        var isEmpty: Bool { low.isEmpty && expired.isEmpty && expiringSoon.isEmpty }
    }

    static func inventoryAlerts(for supplies: [SupplyItem], now: Date = Date()) -> InventorySummary {
        var summary = InventorySummary()
        for item in supplies {
            if isLow(item) { summary.low.append(item) }
            switch expirationStatus(for: item, now: now) {
            case .expired: summary.expired.append(item)
            case .soon: summary.expiringSoon.append(item)
            default: break
            }
        }
        return summary
    }
}

File: expo/utils/supplyAlerts.ts
import { SupplyItem } from '@/types';

const SOON_DAYS = 30;

/**
 * Parses expiration strings like "2026-12", "2026-12-15", or ISO dates.
 * Month-only values are treated as the last day of that month.
 */
export function parseExpirationDate(raw?: string): Date | null {
  if (!raw) return null;
  const trimmed = raw.trim();
  if (!trimmed) return null;

  const monthOnly = /^(\d{4})-(\d{1,2})$/.exec(trimmed);
  if (monthOnly) {
    const year = parseInt(monthOnly[1], 10);
    const month = parseInt(monthOnly[2], 10);
    if (month < 1 || month > 12) return null;
    return new Date(year, month, 0, 23, 59, 59, 999);
  }

  const dayOnly = /^(\d{4})-(\d{1,2})-(\d{1,2})$/.exec(trimmed);
  if (dayOnly) {
    const year = parseInt(dayOnly[1], 10);
    const month = parseInt(dayOnly[2], 10);
    const day = parseInt(dayOnly[3], 10);
    const date = new Date(year, month - 1, day, 23, 59, 59, 999);
    return Number.isNaN(date.getTime()) ? null : date;
  }

  const parsed = new Date(trimmed);
  return Number.isNaN(parsed.getTime()) ? null : parsed;
}

export function isSupplyLow(item: SupplyItem): boolean {
  return item.quantity <= item.minimumQuantity;
}

export function getExpirationStatus(
  item: SupplyItem,
  now: Date = new Date()
): 'expired' | 'soon' | 'ok' | 'none' {
  const date = parseExpirationDate(item.expirationDate);
  if (!date) return 'none';
  if (date.getTime() < now.getTime()) return 'expired';
  const soon = new Date(now);
  soon.setDate(soon.getDate() + SOON_DAYS);
  if (date.getTime() <= soon.getTime()) return 'soon';
  return 'ok';
}

export interface InventoryAlertSummary {
  low: SupplyItem[];
  expired: SupplyItem[];
  expiringSoon: SupplyItem[];
}

export function getInventoryAlerts(
  supplies: SupplyItem[],
  now: Date = new Date()
): InventoryAlertSummary {
  const low: SupplyItem[] = [];
  const expired: SupplyItem[] = [];
  const expiringSoon: SupplyItem[] = [];

  for (const item of supplies) {
    if (isSupplyLow(item)) low.push(item);
    const status = getExpirationStatus(item, now);
    if (status === 'expired') expired.push(item);
    if (status === 'soon') expiringSoon.push(item);
  }

  return { low, expired, expiringSoon };
}
